common/environment/environment.cpp

```diff
@@ -1336,13 +1538,222 @@ void CLocalEnvironment::clearCache()
+
+        SCMStringBuffer dropZoneDir;
+        zoneIt->query().getDirectory(dropZoneDir);
+        const char * pdropzoneDir = dropZoneDir.str();
+        LOG(MCdebugInfo, unknownJob, "Drop zone name:'%s', directory:'%s'", dropZoneName.str(), pdropzoneDir);
+
+        // Check target file path starts with this Drop zone path
+        // the drop zone paths can be nested (nothing forbids it) like
+        // dz1: /a/b/c/d
+        // dz2: /a/b/c
+        // dz3: /a/b
+        // check all and select the shortest
+
+        if ( strncmp(pdropzoneDir, targetFilePath, strlen(pdropzoneDir)) == 0)
+        {
+            if (dropzonePathLen > strlen(pdropzoneDir))
+            {
+                dropzonePathLen = strlen(pdropzoneDir);
+                dropzonePath.set(pdropzoneDir);
+                pdropzonePath = dropzonePath.str();
+                retVal = zoneIt->ptr();
+            }
+        }
+    }
+    return retVal;
+}
+
+IConstDropZoneInfoIterator * CLocalEnvironment::getDropZoneIteratorByComputer(const char * computer) const
+{
+    return (IConstDropZoneInfoIterator *) new CConstDropZoneIteratorByComputer(computer);
+}
+
+IConstDropZoneInfoIterator * CLocalEnvironment::getDropZoneIterator() const
+{
+    return (IConstDropZoneInfoIterator *) new CConstDropZoneInfoIterator();
+}
+
+IConstMachineInfoIterator * CLocalEnvironment::getMachineIterator() const
+{
+    return (IConstMachineInfoIterator *) new CConstMachineInfoIterator();
+}
+
+//==========================================================================================
+// Iterators implementation
+
+CConstMachineInfoIterator::CConstMachineInfoIterator()
+{
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    constEnv.setown((CLocalEnvironment *)factory->openEnvironment());
+    index = 1;
+    maxIndex = constEnv->getNumberOfMachines();
+    curr = NULL;
+    if (maxIndex)
+        curr = constEnv->getMachineByIndex(index);
+}
+
+bool CConstMachineInfoIterator::first()
+{
+    return ((index == 1) && (curr != NULL));
+}
+bool CConstMachineInfoIterator::next()
+{
+    if (index < maxIndex)
+    {
+        index++;
+        curr = constEnv->getMachineByIndex(index);
+    }
+    else
+        curr = NULL;
+
+    return (curr ? true : false);
+}
+
+bool CConstMachineInfoIterator::isValid()
+{
+    return curr != NULL;
+}
+
+IConstMachineInfo & CConstMachineInfoIterator::query()
+{
+    return *curr;
+}
+
+unsigned CConstMachineInfoIterator::count()
+{
+    return maxIndex;
+}
+
+//--------------------------------------------------
+
+CConstDropZoneIteratorByComputer::CConstDropZoneIteratorByComputer(const char * computer)
+{
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    constEnv.setown((CLocalEnvironment *)factory->openEnvironment());
+    computerName.set(computer);
+    maxIndex = constEnv->getNumberOfDropZonesByComputer(computer);
+    index = 1;
+    curr = constEnv->getDropZoneByComputerByIndex(computer, index);
```

Comment:
curr should assigned in first() (so reset when first() is called).
curr should be set to NULL here.